Translate COALESCE as ISNULL

ranma42 opened the pull request:
`COALESCE`is a syntactic shortcut for the CASE expression. As such, the input values are evaluated multiple times.

`ISNULL` does not have this shortcoming.

Fixes #32519.

ranma42 commented:
At least a test for the type propagation is needed

ranma42 commented:
It looks like the type mapping computed in `Coalesce` does not match the actual expression type :(
At least the test was helpful in identifying this issue 😅 
This problem could be related to #14719

raymens commented:
Another application for this is that COALESCE cannot be used for *Indexed Views* in SQL Server, however ISNULL is supported in case of aggregations.

This makes it difficult/impossible to make EF Core generated queries use the indexed views as I suppose the query processor thinks they're different. 

For example:
```
CREATE VIEW [dbo].[vAggregation]
   WITH SCHEMABINDING
   AS
      SELECT a, SUM(ISNULL(b, 0.0)) AS bSum, COUNT_BIG(*) AS Count 
      FROM dbo.c
      GROUP BY a
GO

CREATE UNIQUE CLUSTERED INDEX IDX_vAggregation
   ON vAggregation(a)
GO

SELECT SUM(b) FROM c -- SLOW: not using the indexed view
SELECT SUM(COALESCE(b, 0)) FROM c -- SLOW: not using the indexed view
SELECT SUM(ISNULL(b, 0)) FROM c -- FAST: uses the indexed view
```

cincuranet commented:
@ranma42, gentle nudge on this. We like the `ISNULL` idea and we'd like to see this done.

ranma42 commented:
@cincuranet I will rebase the PR and check the new test results ASAP (worst case: next weekend)

Last time I kind of got stuck because of some typing issues as mentioned in https://github.com/dotnet/efcore/pull/34171#issuecomment-2211724474

I will try to point out explicitly the issues that must be solved (ideally as questions) to unblock this

ranma42 commented on test/EFCore.Specification.Tests/Query/NorthwindMiscellaneousQueryTestBase.cs:
```diff
@@ -723,6 +723,13 @@ public virtual Task Ternary_should_not_evaluate_both_sides_with_parameter(bool a
+    public virtual Task Coalesce_Correct_Type(bool async)
+        => AssertQuery(
+            async,
+            ss => ss.Set<Customer>().Select(c => c.Region ?? "no region specified"));
```
This test is designed to ensure that EFCore takes care of the difference in type propagation between `ISNULL` and `COALESCE`.

From the [docs](https://learn.microsoft.com/en-us/sql/t-sql/language-elements/coalesce-transact-sql?view=sql-server-ver16#comparing-coalesce-and-isnull):
> Data type determination of the resulting expression is different. ISNULL uses the data type of the first parameter, COALESCE follows the CASE expression rules and returns the data type of value with the highest precedence.

Specifically, in this case `c.Region` is an `nvarchar(15)`, while `"no region specified"` cannot be represented as a value of that type (it's longer than 15).

The implementation I did in this PR would take care of this, under the assumption that the `typeMapping` for `CONVERT` can be used to determine the type (and/or whether any casting/conversion is needed), but unfortunately EFCore currently computes it as `nvarchar(15)` even though (I believe) its result is actually an `nvarchar(18)`.

See https://dbfiddle.uk/ITGS6ZVJ

```sql
DECLARE @what sql_variant;
DECLARE @foo varchar(15) = NULL;

SELECT @what = ISNULL(@foo, 'no region selected');
SELECT
    @what,
    SQL_VARIANT_PROPERTY(@what, 'BaseType'),
    SQL_VARIANT_PROPERTY(@what, 'MaxLength');

SELECT @what = COALESCE(@foo, 'no region selected');
SELECT
    @what,
    SQL_VARIANT_PROPERTY(@what, 'BaseType'),
    SQL_VARIANT_PROPERTY(@what, 'MaxLength');
```

ranma42 commented:
What is the best way forward for the type issue?
Also, what should be the scope? Just the SqlServer provider or EFCore in general?

I see that the type inference already handles string concatenation in a special way, but it looks like nothing at all happens for other values/operations (`CASE` and `COALESCE` included), which is kind of suspicious (what should the typeMapping of `c.IntColumn + c.DoubleColumn` be?)

NB: I am not sure I actually grasp what typeMappings are supposed to represent (hence how they should be computed/used), so maybe I am simply misusing it in this PR 😇

cincuranet commented:
> What is the best way forward for the type issue?

Are you talking about the `nvarchar(15)` and `nvarchar(18)`?

> Also, what should be the scope? Just the SqlServer provider or EFCore in general?

Likely just SQL Server, given the `ISNULL` is SQL Server specific.

> (what should the typeMapping of c.IntColumn + c.DoubleColumn be?)

It should result in _double_, I believe.

ranma42 commented:
> If `a` has a type mapping, the type mapping of `a + b`, `a - b`, ... is simply the type mapping of `a`, even if `a` is an `int` and `b` is a `double`.

In most cases this is not actually a problem, as the compiler inserts casts to ensure that the type of `a` and `b` is matching.

ranma42 commented:
I fixed the conflicts and force-pushed the branch. I also added the test for `int`/`double` coalescing, but I found out that it also fails regardless of `ISNULL`, so I opened https://github.com/dotnet/efcore/issues/35516 to track it.

ranma42 commented:
I might need some help for Cosmos 😇

cincuranet commented:
@ranma42 Here's what's failing. The first two seem to be easy renames. Third one is new test, below is what was produced, validate that and change `AssertSql` as needed.

```text
    Microsoft.EntityFrameworkCore.Query.NorthwindMiscellaneousQueryCosmosTest.SelectMany_primitive_select_subquery(async: False) [FAIL]
      Assert.Equal() Failure: Strings differ
                                        ↓ (pos 189)
      Expected: ···"+<>c__DisplayClass172_0[Microsoft.EntityF"···
      Actual:   ···"+<>c__DisplayClass175_0[Microsoft.EntityF"···
                                        ↑ (pos 189)
      Stack Trace:
        D:\a\efcore\efcore\test\EFCore.Cosmos.FunctionalTests\Query\NorthwindMiscellaneousQueryCosmosTest.cs(3380,0): at Microsoft.EntityFrameworkCore.Query.NorthwindMiscellaneousQueryCosmosTest.SelectMany_primitive_select_subquery(Boolean async)
        --- End of stack trace from previous location ---
      Output:
        Compiling query expression: 
        'DbSet<Employee>()
            .Any()'
        
        Generated query execution expression: 
        'queryContext => new QueryingEnumerable<bool>(
            (CosmosQueryContext)queryContext, 
            SqlExpressionFactory, 
            QuerySqlGeneratorFactory, 
            Projection Mapping:
                EmptyProjectionMember -> 0
            SELECT VALUE EXISTS (
                SELECT 1
                FROM root AS c) AS c, 
            Func<QueryContext, JToken, bool>, 
            TestModels.Northwind.NorthwindContext, 
            EntityType: Employee, 
            List<Expression> {  }, 
            False, 
            True
        )
            .Single()'
        
        An exception occurred while iterating over the results of a query for context type 'Microsoft.EntityFrameworkCore.TestModels.Northwind.NorthwindContext'.
        System.InvalidOperationException: An error was generated for warning 'Microsoft.EntityFrameworkCore.Database.SyncNotSupported': Azure Cosmos DB does not support synchronous I/O. Make sure to use and correctly await only async methods when using Entity Framework Core to access Azure Cosmos DB. See https://aka.ms/ef-cosmos-nosync for more information. This exception can be suppressed or logged by passing event ID 'CosmosEventId.SyncNotSupported' to the 'ConfigureWarnings' method in 'DbContext.OnConfiguring' or 'AddDbContext'.
           at Microsoft.EntityFrameworkCore.Diagnostics.EventDefinition.Log[TLoggerCategory](IDiagnosticsLogger`1 logger, Exception exception) in D:\a\efcore\efcore\src\EFCore\Diagnostics\EventDefinition.cs:line 66
           at Microsoft.EntityFrameworkCore.Cosmos.Diagnostics.Internal.CosmosLoggerExtensions.SyncNotSupported(IDiagnosticsLogger`1 diagnostics) in D:\a\efcore\efcore\src\EFCore.Cosmos\Diagnostics\Internal\CosmosLoggerExtensions.cs:line 35
           at Microsoft.EntityFrameworkCore.Cosmos.Storage.Internal.CosmosClientWrapper.ExecuteSqlQuery(String containerId, PartitionKey partitionKeyValue, CosmosSqlQuery query) in D:\a\efcore\efcore\src\EFCore.Cosmos\Storage\Internal\CosmosClientWrapper.cs:line 607
           at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosShapedQueryCompilingExpressionVisitor.QueryingEnumerable`1.Enumerator.MoveNext() in D:\a\efcore\efcore\src\EFCore.Cosmos\Query\Internal\CosmosShapedQueryCompilingExpressionVisitor.QueryingEnumerable.cs:line 151
        
    Microsoft.EntityFrameworkCore.Query.NorthwindMiscellaneousQueryCosmosTest.SelectMany_primitive_select_subquery(async: True) [FAIL]
      Assert.Equal() Failure: Strings differ
                                        ↓ (pos 189)
      Expected: ···"+<>c__DisplayClass172_0[Microsoft.EntityF"···
      Actual:   ···"+<>c__DisplayClass175_0[Microsoft.EntityF"···
                                        ↑ (pos 189)
      Stack Trace:
        D:\a\efcore\efcore\test\EFCore.Cosmos.FunctionalTests\Query\NorthwindMiscellaneousQueryCosmosTest.cs(3380,0): at Microsoft.EntityFrameworkCore.Query.NorthwindMiscellaneousQueryCosmosTest.SelectMany_primitive_select_subquery(Boolean async)
        --- End of stack trace from previous location ---
      Output:
        An exception occurred while iterating over the results of a query for context type 'Microsoft.EntityFrameworkCore.TestModels.Northwind.NorthwindContext'.
        System.InvalidOperationException: An error was generated for warning 'Microsoft.EntityFrameworkCore.Database.SyncNotSupported': Azure Cosmos DB does not support synchronous I/O. Make sure to use and correctly await only async methods when using Entity Framework Core to access Azure Cosmos DB. See https://aka.ms/ef-cosmos-nosync for more information. This exception can be suppressed or logged by passing event ID 'CosmosEventId.SyncNotSupported' to the 'ConfigureWarnings' method in 'DbContext.OnConfiguring' or 'AddDbContext'.
           at Microsoft.EntityFrameworkCore.Diagnostics.EventDefinition.Log[TLoggerCategory](IDiagnosticsLogger`1 logger, Exception exception) in D:\a\efcore\efcore\src\EFCore\Diagnostics\EventDefinition.cs:line 66
           at Microsoft.EntityFrameworkCore.Cosmos.Diagnostics.Internal.CosmosLoggerExtensions.SyncNotSupported(IDiagnosticsLogger`1 diagnostics) in D:\a\efcore\efcore\src\EFCore.Cosmos\Diagnostics\Internal\CosmosLoggerExtensions.cs:line 35
           at Microsoft.EntityFrameworkCore.Cosmos.Storage.Internal.CosmosClientWrapper.ExecuteSqlQuery(String containerId, PartitionKey partitionKeyValue, CosmosSqlQuery query) in D:\a\efcore\efcore\src\EFCore.Cosmos\Storage\Internal\CosmosClientWrapper.cs:line 607
           at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosShapedQueryCompilingExpressionVisitor.QueryingEnumerable`1.Enumerator.MoveNext() in D:\a\efcore\efcore\src\EFCore.Cosmos\Query\Internal\CosmosShapedQueryCompilingExpressionVisitor.QueryingEnumerable.cs:line 151
        
    Microsoft.EntityFrameworkCore.Query.NorthwindMiscellaneousQueryCosmosTest.Coalesce_Correct_Multiple_Same_TypeMapping(async: True) [FAIL]
      Assert.Empty() Failure: Collection was not empty
      Collection: ["SELECT VALUE\r\n{\r\n    \"ReportsTo\" : c[\"Repor"···]
      Stack Trace:
        D:\a\efcore\efcore\test\EFCore.Cosmos.FunctionalTests\TestUtilities\TestSqlLoggerFactory.cs(53,0): at Microsoft.EntityFrameworkCore.TestUtilities.TestSqlLoggerFactory.AssertBaseline(String[] expected, Boolean assertOrder)
        D:\a\efcore\efcore\test\EFCore.Cosmos.FunctionalTests\Query\NorthwindMiscellaneousQueryCosmosTest.cs(5028,0): at Microsoft.EntityFrameworkCore.Query.NorthwindMiscellaneousQueryCosmosTest.AssertSql(String[] expected)
        D:\a\efcore\efcore\test\EFCore.Cosmos.FunctionalTests\Query\NorthwindMiscellaneousQueryCosmosTest.cs(3238,0): at Microsoft.EntityFrameworkCore.Query.NorthwindMiscellaneousQueryCosmosTest.<>c__DisplayClass312_0.<<Coalesce_Correct_Multiple_Same_TypeMapping>b__0>d.MoveNext()
        --- End of stack trace from previous location ---
        D:\a\efcore\efcore\test\EFCore.Cosmos.FunctionalTests\TestUtilities\CosmosTestHelpers.cs(47,0): at Microsoft.EntityFrameworkCore.TestUtilities.CosmosTestHelpers.NoSyncTest(Boolean async, Func`2 testCode)
        --- End of stack trace from previous location ---
      Output:
        ---- New Baseline -------------------------------------------------------------------
                AssertSql(
        """
        SELECT VALUE
        {
            "ReportsTo" : c["ReportsTo"],
            "c" : 1,
            "c0" : 2,
            "c1" : 3
        }
        FROM root c
        ORDER BY c["EmployeeID"]
        """);

```

cincuranet commented on test/EFCore.SqlServer.FunctionalTests/Query/NorthwindMiscellaneousQuerySqlServerTest.cs:
```diff
@@ -2844,6 +2844,41 @@ ELSE [c].[Region]
+
+        AssertSql(
+            """
+SELECT COALESCE([c].[Region], N'no region specified')
```
This could use the `ISNULL` as well, isn't it?

ranma42 replied:
Ideally, yes, but in order to do that we would first need to extend EFCore to correctly compute the type of the expressions as mentioned in my previous comment.

Specifically `SELECT ISNULL([c].[Region], N'no region specified')` would result in the literal truncated to `'no region selec'`.

Instead an appropriate translation would be `SELECT ISNULL(CONVERT(varchar, [c].[Region], N'no region specified'))` or, even better, `SELECT ISNULL(CONVERT(varchar(18), [c].[Region], N'no region specified'))`, but the type that can be observed from the expressions does not take this into account.

cincuranet replied:
Right. I completely forgot [about the](https://learn.microsoft.com/en-us/sql/t-sql/functions/isnull-transact-sql?view=sql-server-ver16#return-types) _Returns the same type as check_expression._.

ranma42 commented on src/EFCore.SqlServer/Query/Internal/SqlServerQuerySqlGenerator.cs:
```diff
@@ -202,6 +202,53 @@ protected override Expression VisitValues(ValuesExpression valuesExpression)
+            var typeMapping = sqlFunctionExpression.TypeMapping;
+            var defaultTypeMapping = _typeMappingSource.FindMapping(type);
+
+            // ISNULL always return a value having the same type as its first
+            // argument. Ideally we would convert the argument to have the
+            // desired type and type mapping, but currently EFCore has some
+            // trouble in computing types of non-homogeneous expressions
+            // (tracked in https://github.com/dotnet/efcore/issues/15586). To
+            // stay on the safe side we only use ISNULL if:
+            //  - all sub-expressions have the same type as the expression
+            //  - all sub-expressions have the same type mapping as the expression
+            //  - the expression is using the default type mapping (combined
+            //    with the two above, this implies that all of the expressions
+            //    are using the default type mapping of the type)
```
@cincuranet I tried to document the current limitation here

roji commented on src/EFCore.SqlServer/Query/Internal/SqlServerQuerySqlGenerator.cs:
```diff
@@ -202,6 +202,53 @@ protected override Expression VisitValues(ValuesExpression valuesExpression)
+            //    with the two above, this implies that all of the expressions
+            //    are using the default type mapping of the type)
+            if (defaultTypeMapping == typeMapping
+                && sqlFunctionExpression.Arguments.All(a => a.Type == type && a.TypeMapping == typeMapping)) {
```
Related to the above comment on the CLR type, I think we could compare the store type instead of referencing-comparing the type mapping instances... There's no guarantee in EF that there's only one type mapping instance for a given type mapping, and unless I'm mistaken the only thing that matters here for ISNULL is the actual store type. So comparing the store type should be safe and possibly allow for some more scenarios.

```suggestion
                && sqlFunctionExpression.Arguments.All(a => a.Type == type && a.TypeMapping?.StoreType == typeMapping?.StoreType)) {
```

ranma42 replied:
IIRC this would not recognize cases in which the mismatch is in one of the parameters (such as `varchar` vs `varchar(15)`).
Maybe the right approach here would be to make typemappings `IEquatable` and compare them

roji replied:
> IIRC this would not recognize cases in which the mismatch is in one of the parameters (such as varchar vs varchar(15)).

I think StoreType must always be the full store type - including any facets. If that's not the case, we most probably have bugs somewhere else as well. Do you have a specific case in mind where this doesn't work?

> Maybe the right approach here would be to make typemappings IEquatable and compare them

I'm not sure... The point here is that in this specific context, the only thing we care about is the actual type in the database; in other words, various other details that distinguish type mappings from one another are irrelevant (value comparers, value converters...), since they affect EF client-side processing only. Assuming we have a 100% unambiguous representation of the database type - which is what StoreType is supposed to be - comparing those should be sufficient and allow for maximum coverage of this optimization, I think.

A general IEquatable would need to also compare e.g. value comparers (since it's possible that in other contexts that's relevant), and so would needlessly exclude some valid cases...

roji commented on src/EFCore.SqlServer/Query/Internal/SqlServerQuerySqlGenerator.cs:
```diff
@@ -202,6 +202,53 @@ protected override Expression VisitValues(ValuesExpression valuesExpression)
+                    .Aggregate(head, (l, r) => new SqlFunctionExpression(
+                        "ISNULL",
+                        arguments: [l, r],
+                        nullable: true,
```
At least for the top-most ISNULL, shouldn't we be taking the original coalesce expression's nullability? For example, in the very common case of `b.Foo ?? "<unknown>"`, the entire expression should be non-nullable since the last sub-expression (the constant string) is non-nullable.

As this is done extremely late in the query pipeline (SQL generator), there's not going to be anything afterwards that cares about this (this node is immediately processed, generating SQL string data, and then discarded). But for correctness's sake (and in case this code is copy-pasted somewhere else), maybe we should do it?

ranma42 replied:
`nullable: true` is always safer than `nullable: false`, so no risk of incorrectness here ;)
We could do this for the top-most `ISNULL` invocation (which, if the `COALESCE` has been optimized properly, is the only one that can be non-nullable), but it would make the code a little more complicated for no apparent advantage.

Note that in this context we do not have access to the results of the nullability processor (the `nullable` value in the expression does not correspond to the results of the processing).

roji replied:
> no risk of incorrectness here

Right, not incorrect in the sense of incorrect results - just potentially SQL that's more convoluted (and less efficient) than it needs to be.

> Note that in this context we do not have access to the results of the nullability processor (the nullable value in the expression does not correspond to the results of the processing).

That's true, but the expression `nullable` does represent what it represents: it seems odd to basically lose that setting just because we're transforming one SQL function into another here.

But anyway, given where this code is, I agree it's theoretical.

ranma42 replied:
> > no risk of incorrectness here
> 
> Right, not incorrect in the sense of incorrect results - just potentially SQL that's more convoluted (and less efficient) than it needs to be.
> 
> > Note that in this context we do not have access to the results of the nullability processor (the nullable value in the expression does not correspond to the results of the processing).
> 
> That's true, but the expression `nullable` does represent what it represents: it seems odd to basically lose that setting just because we're transforming one SQL function into another here.

I believe we are not losing any information here: `COALESCE` expressions are always constructed with `nullable: true` from [this code](https://github.com/dotnet/efcore/blob/main/src/EFCore.Relational/Query/SqlExpressionFactory.cs#L592-L599).
They can be re-written in a few places but the `IsNullable` property is never modified (neither for `COALESCE` nor for other function expressions).

> 
> But anyway, given where this code is, I agree it's theoretical. If we do any extra work on this, I'd add a comment making that clear.

roji replied:
I see, I wasn't aware of that... That's presumably because we have special handling of COALESCE in any case in SqlNullabilityProcessor, so IsNullable is maybe effectively unused?

I'd prefer it IsNullable were accurate here (across the board, regardless of COALESCE vs. ISNULL) - you never know when someone might look at it - but I guess it's not super important at this point.

ranma42 replied:
> I see, I wasn't aware of that... That's presumably because we have special handling of COALESCE in any case in SqlNullabilityProcessor, so IsNullable is maybe effectively unused?

The nullability processor relies on `IsNullable` as input to decide whether to assume that the function is guaranteed to return a non-null value or whether its result should be considered (non)nullable depending on (some of) its arguments.
Currently it is accurate in the perspective of function definition and not from the point of view of the expression (even less of the expression in a given context).
IIRC the same holds true for other expressions as well: during the nullability computation for `myNullableColumn is null ? 1 : myNullableColumn + 2` the second `myNullableColumn` expression is evaluated as non-nullable, but it is not rewritten to have `IsNullable = false`.

> I'd prefer it IsNullable were accurate here (across the board, regardless of COALESCE vs. ISNULL) - you never know when someone might look at it - but I guess it's not super important at this point.

Your suggestion goes in the direction of tracking nullability in the expression, which might indeed prove very useful, for example when lowering SqlServer boolean expressions/predicates 😫, but I think is not specifically related to the change COALESCE->ISNULL in this PR

roji commented on src/EFCore.SqlServer/Query/Internal/SqlServerQuerySqlGenerator.cs:
```diff
@@ -202,6 +202,53 @@ protected override Expression VisitValues(ValuesExpression valuesExpression)
+            // stay on the safe side we only use ISNULL if:
+            //  - all sub-expressions have the same type as the expression
+            //  - all sub-expressions have the same type mapping as the expression
+            //  - the expression is using the default type mapping (combined
```
This one also seems a bit strict - if all sub-expressions have the same store type, shouldn't everything be fine and we can assume the overall expression has the same type? What's the scenario that could be problematic given a non-default store type?

ranma42 replied:
EFCore computes accurate store types only in some cases (column expressions, simple additions), but in many other cases it approximates them, for example a literal string is modeled as `varchar` instead of `varchar(length-of-the-literal)`.

The idea was that if all of the the expressions are using the default type mapping, the C#-oriented type compatibility matches the SQL one as closely as possible... but now I wonder if some cases are actually unsafe even under this assumption 🤔 .
I think an example that would break with only same type and same type-mapping is:
```csharp
(col1VarChar10 != 'foo' ? col1VarChar10 : 'a long literal') ??
(col2VarChar10 != 'foo' ? col2VarChar10 : 'a very long literal')
```

IIRC EFCore computes the type mapping of the LHS (and of the whole expression) as the same type mapping as `col1VarChar10`.

I think an example that would break with the current set of constraints is :
```csharp
(col1VarChar10 == 'foo' ? 'a long literal' : col1VarChar10) ??
(col2VarChar10 == 'foo' ? 'a very long literal' : col2VarChar10)
```

I will try to check this ASAP.

ranma42 replied:
`Coalesce_Correct_TypeMapping_Double` and `Coalesce_Correct_TypeMapping_String` are a couple of examples in which EFCore computes the wrong (/inaccurate) type mapping for some of the expressions.
Assuming that the implementation in this PR is wrong, I will add another test case that shows that the check is not strict enough.

Conversely, you are right that checking that the type is the default one is not required; its main purpose was to trust C# for the type computation, under the (presumably wrong) assumption that when the C#-computed type and the EFCore-computed type matched, that also meant a match for the SQL type.

roji reviewed (commented):
Sorry for the post-merge review... It's really nice to see this merged - should provide a nice perf improvement when coalescing heavy expressions, and provide sane behavior when coalescing non-deterministic ones.

See below for some minor thoughts that came up (nothing crucial).

ranma42 commented:
This test passes with the current code but would fail if the check for the default type(mapping) was removed:

```csharp
    [ConditionalTheory]
    [MemberData(nameof(IsAsyncData))]
    public virtual Task Coalesce_Correct_TypeMapping_String_Expression(bool async)
        => AssertQuery(
            async,
            ss => ss.Set<Customer>().OrderBy(c => c.CustomerID).Select(c => (c.Region == null ? null : "R" + c.Region) ?? "no region specified"),
            assertOrder: true);
```

I am now pretty confident I can make the current translation fail with a different test (still working on it, though).

(also, the `StoreType` comparison is probably good; while experimenting I recalled I was trying to avoid the cases which involve type conversion, but iiuc the current state of EFCore is that those are known to behave in weird ways when used in expressions)

roji commented:
> iiuc the current state of EFCore is that those are known to behave in weird ways when used in expressions

Yes, that's true.. I don't think we need to take special account of this for this particular case.

Thanks for doing the extra checks around the default type mapping...
